src/ShallowWaterExplicit.jl

```diff
@@ -0,0 +1,171 @@
+function shallow_water_explicit_time_step!(model, dΩ, dω, f, g, h₁, u₁, hₘ, uₘ, hₚ, uₚ, RTMMchol, L2MMchol, dt, leap_frog, τ, Q, V, R, S, h₂, u₂, ϕ, F)
+  # energetically balanced explicit second order shallow water solver
+  # reference: eqns (21-24) of
+  # https://github.com/BOM-Monash-Collaborations/articles/blob/main/energetically_balanced_time_integration/EnergeticallyBalancedTimeIntegration_SW.tex
+  #
+  # f          : coriolis force (field)
+  # g          : gravity (constant)
+  # h₁         : fluid depth at current time level
+  # u₁         : fluid velocity at current time level
+  # hₘ         : fluid depth at previous time level (for leap-frogging the first step)
+  # uₘ         : fluid velocity at previous time level (for leap-frogging the first step)
+  # RTMM       : H(div) mass matrix, ∫β⋅βdΩ, ∀β∈ H(div,Ω)
+  # L2MM       : L² mass matrix, ∫γγdΩ, ∀γ∈ L²(Ω)
+  # dt         : time step
+  # leap_frog  : do leap frog time integration for the first step (boolean)
+  # dΩ         : measure of the elements
+
```

Comment:
I am still concerned about the performance of this function, i.e., `shallow_water_explicit_time_step!`. I think it can still be significanlty improved. In particular, as the mesh does not change along the simulation (e.g., we do not currently have dynamic mesh adaptation),  the size of matrices and vectors, and the symbolic sparsity pattern of matrices remains the same.  Thus, there is a lot of potential for memory to be reused among calls to this function. Currently we are dynamically allocating at each call:

1. Eight vectors, namely `rhs1`-`rhs8`.
2. The two coefficient matrices of the two potential vorticity linear systems. (steps 1.3 and 2.3).
3. The two LU factorizations  of the coefficient matrices in 2. Note that, although we need to compute these LU factorizations at each step (as the entries of the matrices in 2. change), we may try to reuse the symbolic sparsity pattern of the L and U factors, and the memory buffers of the L and U factors. 

At present, I know how to address 1. and 2. I think that 3. can also be addressed using low level Gridap functions. Let us try to first address 1. and 2. in a different comment box.

Reply:
In order to address 1. and 2. we can use in-place `assemble_vector!` and `assemble_matrix_and_vector!` assembly functions.  

The idea is that we pass work array vectors (do we actually need eight `rhs` vectors?) and work matrices (do we actually need to matrices or just a single one?)  to `shallow_water_explicit_time_step!` and we call these functions to perform assembly in-place instead of generating whole new objects each time. Note that generating a sparse matrix from scratch is much more expensive than reusing an already generated matrix, as with the former we need to compute also the sparsity pattern, i.e., where the nonzero entries are located. 

It turns out that the functions that are more appropiate in your particular context are not already available in the latest version of Gridap. However, I have developed them in the following PR:  https://github.com/gridap/Gridap.jl/pull/652, which is currently under review and will be merged soon. See, e.g., 

https://github.com/gridap/Gridap.jl/blob/developed_additional_highlevel_assembly_functions/test/FESpacesTests/AssemblersTests.jl#L78

for examples of usage of these functions.

Finally, in order to help you out, I have pulled all these changes into the Gridap.jl branch your are pointing to from `GridapGeosiences.jl`. Thus, if you just run `up Gridap` from the package manager, you will have these two functions are your disposal.

Reply:
In order to address 3., you can look at lines 246-247 of the following source file of `Gridap.jl` 

https://github.com/gridap/Gridap.jl/blob/master/src/Algebra/LinearSolvers.jl#L249

In particular, we are interested into calling the `numerical_setup!` and `solve!` functions inside the function. 

A short snippet is as follows (where `A` is assumed to be an assembled sparse matrix):

```julia
ls = LUSolver()
lss = symbolic_setup(ls,A)
lsn = numerical_setup(lss,A) 
solve!(x,lsn,b)                         # Solve A x = b 
                                                   # Change the entries of `A` by whatever process ...
numerical_setup!(lsn,A)     #  Reucse L,U memory buffers to recompute LU factorization in-place
solve!(x,lsn,b)                         # Solve Ax = b with the updated A
```

Let me know if you face issues with this ...

Reply:
Ok I see how you have solved this comment! Good approach!